File: chapter03/cpp/calculator.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdlib.h>
#include <stack>

using namespace std;

int calculate(const string& s)
{
    stack<int> st;
    int num = 0;
    char sign = '+';
    int res = 0;
    for(int i=0; i<s.size(); i++)
    {      
        int tmp;
        char c = s[i];
        if(isdigit(c))
            num = 10 * num + (c - '0');
        
        // 第二个条件是保证把最后一个整数入栈
        if(!isdigit(c) && c != ' ' || i == s.size() - 1)
        {
            switch (sign) {
                case '+':
                    st.push(num); break;
                case '-':
                    st.push(-num); break;
                case '*':
                    tmp = st.top() * num;
                    st.pop();
                    st.push(tmp);
                    break;
                case '/':
                    tmp = st.top() / num;
                    st.pop();
                    st.push(tmp);
                    break;
            }
            num = 0;
            sign = c;
        }
    }
    while (st.size() != 0)
    {
        res += st.top();
        st.pop();
    }
        
    return res;
}

int calculateII(vector<char> v)
{
    char sign = '+';
    int num = 0;
    int res = 0;
    stack<int> st;
    int tmp;
    while(v.size() > 0)
    {
        char c = *v.begin();
        v.erase(v.begin());

        if(isdigit(c))
            num = 10 * num + (c - '0');
        if(!isdigit(c) && c != ' ' || v.size()==0)
        {
            switch (sign) {
                case '+':
                    st.push(num); break;
                case '-':
                    st.push(-num); break;
                case '*':
                    tmp = st.top() * num;
                    st.pop();
                    st.push(tmp);
                    break;
                case '/':
                    tmp = st.top() / num;
                    st.pop();
                    st.push(tmp);
                    break;
            }
            num = 0;
            sign = c;
        }
    }
    while (st.size() != 0)
    {
        res += st.top();
        st.pop();
    }
    return res;
}

// vector可换成list
int calculate_parentheses(vector<char>& v)
{
    char sign = '+';
    int num = 0;
    int res = 0;
    stack<int> st;
    int tmp;
    while(v.size() > 0)
    {
        char c = *v.begin();
        v.erase(v.begin());

        if(isdigit(c))
            num = 10 * num + (c - '0');
        
        if(c == '(')
            num = calculate_parentheses(v);
        
        if((!isdigit(c) && c != ' ') || (v.size()==0 && c != ')'))
        {
            switch (sign) {
                case '+':
                    st.push(num); break;
                case '-':
                    st.push(-num); break;
                case '*':
                    tmp = st.top() * num;
                    st.pop();
                    st.push(tmp);
                    break;
                case '/':
                    tmp = st.top() / num;
                    st.pop();
                    st.push(tmp);
                    break;
            }
            num = 0;
            sign = c;
        }

        if(c == ')')
            break;
    }
    while (st.size() != 0)
    {
        res += st.top();
        st.pop();
    }
    return res;
}

void usage()
{
    string s = "-1-12*(3+1)";
    vector<char> v(s.size());
    for(int i=0; i<s.size(); i++)
        v[i] = s[i];
    cout << s << " = " << calculate_parentheses(v) << endl;
}

int main()
{
    usage();
    return 0;
}

File: chapter03/python/calculator.py
class Calculator:
    def __init__(self):
        self.operator = []
        self.num = []

    def calculate(self, s):
        if s[0].isdigit():
            self.operator.append("+")
        res = 0
        for e in s:
            if e.isdigit():
                if self.operator[-1] == "(":
                    self.operator.append("+")
                    self.num.append(int(e))
                else:
                    self.num.append(int(e))
            elif e == " ":
                continue
            elif e == ")":
                tmp = 0
                while self.operator[-1] != "(":
                    tmp += self.do_operator()
                self.operator.pop()
                self.num.append(tmp)
            else:
                self.operator.append(e)

        while len(self.num) > 0:
            res += self.do_operator()

        return res

    def do_operator(self):
        res = 0
        operator = self.operator.pop()
        if operator == "+":
            res += self.num.pop()
        elif operator == "-":
            res -= self.num.pop()
        elif operator == "*":
            tmp = self.num.pop()
            self.num.append(self.num.pop() * tmp)
        elif operator == "/":
            tmp = self.num.pop()
            self.num.append(self.num.pop() / tmp)
        return res


if __name__ == "__main__":
    s = "3*(-2+(3-2)*(4-1))"

    c = Calculator()
    print(c.calculate(s))

